crates/cadenza-eval/src/eval.rs

```diff
@@ -1270,6 +910,680 @@ pub fn builtin_measure() -> BuiltinMacro {
+                (a, _) => Err(Diagnostic::type_error(
+                    Type::union(vec![Type::Integer, Type::Float]),
+                    a.type_of(),
+                )),
+            }
+        },
+    }
+}
+
+/// Creates the `==` equality operator.
+pub fn builtin_eq() -> BuiltinFn {
+    BuiltinFn {
+        name: "==",
+        signature: Type::function(vec![Type::Unknown, Type::Unknown], Type::Bool),
+        func: |args, _ctx| {
+            if args.len() != 2 {
+                return Err(Diagnostic::arity(2, args.len()));
+            }
+
+            let a = &args[0];
+            let b = &args[1];
+
+            // Check that types match
+            let type_a = a.type_of();
+            let type_b = b.type_of();
+            if type_a != type_b {
+                return Err(Diagnostic::type_error(type_a, type_b));
+            }
+
+            Ok(Value::Bool(a == b))
+        },
+    }
+}
+
+/// Creates the `!=` inequality operator.
+pub fn builtin_ne() -> BuiltinFn {
+    BuiltinFn {
+        name: "!=",
+        signature: Type::function(vec![Type::Unknown, Type::Unknown], Type::Bool),
+        func: |args, _ctx| {
+            if args.len() != 2 {
+                return Err(Diagnostic::arity(2, args.len()));
+            }
+
+            let a = &args[0];
+            let b = &args[1];
+
+            // Check that types match
+            let type_a = a.type_of();
+            let type_b = b.type_of();
+            if type_a != type_b {
+                return Err(Diagnostic::type_error(type_a, type_b));
+            }
+
+            Ok(Value::Bool(a != b))
+        },
+    }
+}
+
+/// Helper function to compare two values using their PartialOrd implementation.
+/// Returns a type error if the values have different types or are not comparable.
+fn compare_ordered<F>(a: &Value, b: &Value, check_ordering: F) -> Result<Value>
+where
+    F: FnOnce(std::cmp::Ordering) -> bool,
+{
+    // Check that types match - strongly typed, no implicit conversions
+    let type_a = a.type_of();
+    let type_b = b.type_of();
+    if type_a != type_b {
+        return Err(Diagnostic::type_error(type_a, type_b));
+    }
+
+    match (a, b) {
+        (Value::Integer(a), Value::Integer(b)) => {
+            let ordering = a.cmp(b);
+            Ok(Value::Bool(check_ordering(ordering)))
+        }
+        (Value::Float(a), Value::Float(b)) => {
+            // For floats, use partial_cmp since they may be NaN
+            match a.partial_cmp(b) {
+                Some(ordering) => Ok(Value::Bool(check_ordering(ordering))),
+                None => Err(Diagnostic::syntax("cannot compare NaN values".to_string())),
+            }
+        }
+        _ => Err(Diagnostic::syntax(format!(
+            "cannot compare values of type {}",
+            a.type_of()
+        ))),
+    }
+}
+
+/// Creates the `<` less-than operator.
+pub fn builtin_lt() -> BuiltinFn {
+    BuiltinFn {
+        name: "<",
+        signature: Type::function(
```

Comment:
Again all of these type signatures need to be unions over the functions, not unions over the arguments